### DISCUSSION TITLE
Client Side Caching

### DISCUSSION THREAD

**Discussion by iwanbk** [redis/go-redis#2370] [General]
Hi,

Is there a plan to add client side caching feature? https://redis.io/docs/manual/client-side-caching/

And how i can help to make it implemented?

Thanks

**Comment by manchurio** [comment]
I also want this feature too.

**Comment by rsanchez-scwx** [comment]
Any updates on this? This feature has been around since Redis v6.0. It's now at v7.x.

**Comment by nicksherron** [comment]
Is there any conversation that I'm missing regarding this?? I too would like this feature and would even be inclined to submit a pr if a maintainer would be open to giving me a rough explanation of the preferred implementation. Thoughts @ofekshenawa ?

**Comment by chayim** [comment]
Folks,  @ofekshenawa has started working on this. In fact we'd definitely welcome community eyes and help - of course! There are currently beta releases available in other cilents, such as [python](https://github.com/redis/redis-py/releases/tag/v5.1.0b3) for people interested in general, in learning.

@uglide Can you share some of your thoughts here?

**Reply by nicksherron** [reply]
Cool thanks for the info!

**Comment by mhelmich** [comment]
Hi y'all, I stumbled across this conversation and put the following PR together. It's not the full-fledged caching mechanism that people are asking for. But it makes a start.
https://github.com/redis/go-redis/pull/3007

**Reply by mhelmich** [reply]
At-mentioning contributors for viz @chayim @ofekshenawa .
It would be nice to get some traction on this PR. Thanks in advance.

**Comment by chisenberg** [comment]
Hello! I also started working on client side caching and had problem receiving the "invalidate" messages from redis.
The pull request [#2370](https://github.com/redis/go-redis/pull/3326) solves this, but you should still find out the CLIENT ID of the subscriber to enable CLIENT TRACKING REDIRECT {{ID}}.

We made the following to work:
- Set the ClientName fileld when creating a new redis.NewClient(...)
- Subscribe to topic "\_\_redis\_\_:invalidate"
- List all pubsub clients on redis with "CLIENT LIST TYPE PUBSUB"
- Filter the ones with the same name used on NewClient
- Get the client id
- Call "CLIENT TRACKING REDIRECT {{ID}} BCAST PREFIX {{optiona prefix}}"

With my pull request it is possible to receive messages with the slice of keys invalidated by other clients or redis.

Hope this helps someone else :smile:

**Comment by nicksherron** [comment]
Imagine the performance and code simplification this feature could bless us all with. I imagine a cache that really only needs `New` and `Get` methods ( likely with some basic fallback expiration ) for high read rarely changed data ( very common use case). 

Makes me wonder why we don’t have any movement on this after over 3  years of asking about it?

**Reply by nicksherron** [reply]
Maybe I spoke too soon, hopefully something like https://github.com/redis/go-redis/pull/3748 gets merged in and makes me eat my words.

**Reply by ndyakov** [reply]
@nicksherron you are correct, we are working on CSC. Maybe you will find this quite interesting as well - https://github.com/redis/go-redis/issues/3883

**Comment by ndyakov** [comment]
Client side caching is not merged on master! https://github.com/redis/go-redis/tree/master/example/client-side-caching
